### PR TITLE
Remove bcrypt dependency, use PBKDF2

### PR DESCRIPTION
**Pull request by justmoon**
Bcrypt has some advantages over PBKDF2 [1], however, PBKDF2 is good enough and built into Node.js, removing a native dependency. PBKDF2 is the algorithm recommended by NIST [2] for password-based key derivation. (We're using it for password hashing which is a different, but related use case.)

This patch also adds some functionality, namely:

* verifyPassword method (required for verifying passwords)
* ability to customize the number of hashing iterations
* automatically reduce number of iterations in testing (NODE_ENV=unit)

[1] http://security.stackexchange.com/questions/4781/do-any-security-experts-recommend-bcrypt-for-password-storage
[2] http://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-132.pdf

### CODE REVIEW

File: utils/hashPassword.js [hunk 1]
```diff
@@ -1,16 +1,132 @@
...
+/**
+ * Verify a password using Node's asynchronous pbkdf2 (key derivation) function.
+ *
+ * Accepts a hash and salt generated by hashPassword, and returns whether the
+ * hash matched the password (as a boolean).
+ *
+ * @param {!String} password
+ * @param {!Buffer} combined Buffer containing hash and salt as generated by
+ *   hashPassword.
+ * @param {Promise<Boolean>} Whether the password matched the hash
```
Comment (sentientwaffle) [inline utils/hashPassword.js]: `@return`, not `@param`

File: utils/hashPassword.js [hunk 2]
```diff
@@ -1,16 +1,132 @@
...
+  const hash = combined.slice(saltBytes + 8)
+
+  return new Promise((resolve, reject) => {
+    // Verify the salt and hash against the password
+    crypto.pbkdf2(password, salt, iterations, hashBytes, PBKDF2_DIGEST, (err, verify) => {
+      if (err) {
+        return reject(err)
+      }
+
+      resolve(Buffer.compare(verify, hash) === 0)
```
Comment (sentientwaffle) [inline utils/hashPassword.js]: `resolve(verify.equals(hash))` would be a little simpler. ([Buffer#equals](https://nodejs.org/api/buffer.html#buffer_buf_equals_otherbuffer))

**Comment by emschwartz** [comment]
LGTM

**Comment by clark800** [comment]
LGTM

**Pull request merged**